File: code/java/src/main/java/cn/az/code/year2020/dec/ValidBst.java
package cn.az.code.year2020.dec;

import cn.az.code.common.TreeNode;

import java.util.Stack;

/**
 * @author az
 * @since 12/16/20 21:19
 */
public class ValidBst {

    public boolean isValidBST(TreeNode root) {
        if (root == null) {
            return true;
        }
        Stack<TreeNode> stack = new Stack<>();
        TreeNode pre = null;
        while (root != null || !stack.isEmpty()) {
            while (root != null) {
                stack.push(root);
                root = root.left;
            }
            root = stack.pop();
            if (pre != null && root.val <= pre.val) {
                return false;
            }
            pre = root;
            root = root.right;
        }
        return true;
    }
}

File: src/tree/valid_bst.py
from index import TreeNode
import sys


def valid_bst(root: TreeNode) -> bool:
    def compare(node: TreeNode, mi, ma) -> bool:
        if not node:
            return True
        if node.val < mi or node.val > ma:
            return False
        l = compare(node.left, mi, node.val)
        r = compare(node.right, node.val, ma)
        return l and r

    return compare(root, -sys.maxsize, sys.maxsize)
